### DISCUSSION TITLE
Rename PhysicalFileSystemProvider to WebRootImageProvider

### DISCUSSION THREAD

**Discussion by ronaldbarendse** [SixLabors/ImageSharp.Web#204] [Ideas] [answered]
The current `PhysicalFileSystemProvider` uses the `IFileProvider` abstraction from `environment.WebRootFileProvider`, which doesn't have to be a physical file system (it's just that ASP.NET Core does use `PhysicalFileProvider` by default).

I would suggest adding a (base?) class that accepts any `IFileProvider` implementation (e.g. `FileProviderImageProvider[Base]`) and rename the current implementation to `WebRootImageProvider`.

The same is true for `PhysicalFileSystemResolver`, which uses the `IFileInfo` abstraction: this could be renamed to `FileInfoImageResolver` (and be re-used by any of the provider implementations).

--------------

It's a bit different for `PhysicalFileSystemCache`, as that does use `System.IO.File` (as it needs write access, which the `IFileProvider` API doesn't provide). So this name makes perfect sense, but it's also using the `PhysicalFileProvider`/`IFileProvider` abstraction (and therefore also requires a dependency on `Microsoft.Extensions.FileProviders.Physical`)! I would suggest removing this and use direct file paths instead...

@JimBobSquarePants Let me know what you think about these (breaking) changes. I'll be happy to create a PR for this 😃

**Comment by JimBobSquarePants** [comment, marked as answer]
Yeah.... The current implementation isn't pretty and we should fix it while we can.

I actually think we should do the opposite and simply mirror the `PhysicalFileSystemCache`. We can provide options containing a `ProviderRootPath` property and default to the `IWebHostEnvironment.WebRootPath` if not set. (I'd probably rename `PhysicalFileSystemCacheOptions.CacheRoot` to `CacheRootPath` also.)

I'm loath to introduce additional base classes. It's an additional opportunity for me to miss something important which makes them unusable.

**Reply by deanmarcussen** [reply]
I agree on that @JimBobSquarePants @ronaldbarendse 

Would actually suggest, since we're at it, that `PhysicalFileSystemCache` should also take a `ProviderRootPath`, and default to `WebRootPath`

I didn't do that last time, but we've had a  couple of requests to move it outside of the webroot, for users that are securing images, and others that can see a use for a cache folder per tenant (makes managing cache sizes/quotas a little more practical).

Not sure about the base classes, we have our own custom implementation, and it wouldn't benefit from them, it's not a lot on the interface to implement.

**Comment by JimBobSquarePants** [comment]
`PhysicalFileSystemCache` already has that but it's called `CacheRoot`

https://github.com/SixLabors/ImageSharp.Web/blob/7c2b4969a7587610ee2e23bcd28e5e74b49ee8d4/src/ImageSharp.Web/Caching/PhysicalFileSystemCacheOptions.cs#L16-L28
